remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -422,75 +428,134 @@ impl REClient {
+                    ));
+                }
+
+                blob_hashes.push(hash);
+                continue;
+            }
+
+            // otherwise, calculate how many messages to send by dividing the size by the limit
+            // we need this to figure out which message is the last one
+            let slop = size % grpc_msg_limit;
+            let num_msgs = (size / grpc_msg_limit) + (if slop > 0 { 1 } else { 0 });
+
+            let vs = stream::iter(data).chunks(grpc_msg_limit as usize);
+            let ws = vs.enumerate().map(move |(i, chunk)| WriteRequest {
+                resource_name: resource_name.clone(),
+                write_offset: (i * grpc_msg_limit as usize) as i64,
+                finish_write: i == (num_msgs as usize - 1),
+                data: chunk.to_vec(),
+            });
+
+            let resp = bs_client.write(ws).await?;
+            if resp.into_inner().committed_size != size as i64 {
+                return Err(anyhow::anyhow!(
+                    "Failed to upload inline blob: invalid committed_size from WriteResponse"
+                ));
+            }
+
+            blob_hashes.push(hash);
+        }
+
+        // upload files, and chunk WriteRequests into small slices too
+        const BUFFER_SIZE: usize = 512 * 1024; // XXX FIXME (aseipp): use a larger dynamic (4MB?) buffer?
+        let mut buffer = [0u8; BUFFER_SIZE];
+
+        let files_with_digest = request.files_with_digest;
+        for file in files_with_digest.unwrap_or_default() {
+            let hash = file.digest.hash;
+            let size = file.digest.size_in_bytes;
+            let uuid = uuid::Uuid::new_v4().to_string();
+            let name = file.name;
+
+            let mut file = tokio::fs::File::open(name).await?;
+            let mut write_offset = 0;
+
+            let mut rs = Vec::new();
+            loop {
+                let resource_name = format!(
+                    "{}uploads/{}/blobs/{}/{}",
+                    instance_prefix,
+                    uuid,
+                    hash.clone(),
+                    size
+                );
+                let n = file.read(&mut buffer).await?;
+                let data = buffer[..n].to_vec();
+                let finish_write = n == 0;
+
+                rs.push(WriteRequest {
+                    resource_name: resource_name.clone(),
+                    write_offset,
+                    finish_write,
+                    data,
+                });
+
+                write_offset += n as i64;
+                if n == 0 {
+                    break;
+                }
+            }
+            let resp = bs_client.write(stream::iter(rs)).await?;
+
+            if resp.into_inner().committed_size != size as i64 {
+                return Err(anyhow::anyhow!(
+                    "Failed to upload file: invalid committed_size from WriteResponse"
+                ));
+            }
+
+            blob_hashes.push(hash);
```

Comment:
Note: this buffers the whole file in memory, which isn't worse than what was here before, but perhaps could be improved (no big deal if you'd rather not).